File: src/simulation/protocol.ts
export const MAX_FRAME_BYTES = 1_048_576;
export const MAX_NEURAL_UPDATES = 20_000;
export const MAX_STRING_LENGTH = 256;
export const MAX_SEQUENCE = 2 ** 53 - 1;

export type PopulationSize = 80 | 1000 | 5000 | 20000 | 124289;
export type BackendPreference = 'auto' | 'cpu' | 'gpu' | 'gpu-strict';
export type Action = 'forward' | 'backward' | 'left' | 'right' | 'wait';
export type KeyName = 'W' | 'A' | 'S' | 'D' | 'SPACE_LEFT' | 'SPACE_RIGHT';
export type MotorPhase = 'neutral' | 'targeting' | 'reaching' | 'pressing' | 'confirmed' | 'retracting' | 'settling' | 'failed';

type Envelope = {
  version: 2;
  sessionId: string;
  episodeId: string;
  sequence: number;
  simulationTime: number;
};

export type Ready = Envelope & { type: 'ready'; backend: 'cpu' | 'gpu'; population: PopulationSize; graphHash: string; checkpointHash: string; acceptedVersions?: 2[]; maxFrameBytes?: number; maxNeuralUpdates?: number };
export type ResetComplete = Envelope & { type: 'reset_complete' };
export type Intention = Envelope & { type: 'intention'; intentionId: string; action: Action; motorPhase: MotorPhase };
export type Snapshot = Envelope & { type: 'snapshot'; body: [number, number, number]; joints: number[]; keys: Partial<Record<KeyName, number>> };
export type NeuralKeyframe = Envelope & { type: 'neural_keyframe'; updates: NeuralUpdate[] };
export type NeuralDelta = Envelope & { type: 'neural_delta'; updates: NeuralUpdate[] };
export type Contact = Envelope & { type: 'contact'; intentionId: string; requestedKey: KeyName; touchedKey: KeyName | null; travel: number; force: number; debounce: number; confirmed: boolean };
export type ActionResult = Envelope & { type: 'action_result'; intentionId: string; result: 'confirmed' | 'waited' | 'failed' };
export type Metrics = Envelope & { type: 'metrics'; physicsHz: number; motorHz: number; neuralHz: number; renderHz: number; latencyMs: number; droppedRenderFrames: number; backendUtilization?: number };
export type Paused = Envelope & { type: 'paused'; code: string; message: string };
export type ErrorMessage = Envelope & { type: 'error'; code: string; message: string };
export type NeuralUpdate = { neuronId: number; value: number };
export type ServerMessage = Ready | ResetComplete | Intention | Snapshot | NeuralKeyframe | NeuralDelta | Contact | ActionResult | Metrics | Paused | ErrorMessage;

const populations = new Set<PopulationSize>([80, 1000, 5000, 20000, 124289]);
const resolvedBackends = new Set(['cpu', 'gpu']);
const actions = new Set<Action>(['forward', 'backward', 'left', 'right', 'wait']);
const keyNames = new Set<KeyName>(['W', 'A', 'S', 'D', 'SPACE_LEFT', 'SPACE_RIGHT']);
const motorPhases = new Set<MotorPhase>(['neutral', 'targeting', 'reaching', 'pressing', 'confirmed', 'retracting', 'settling', 'failed']);
const serverTypes = new Set<ServerMessage['type']>(['ready', 'reset_complete', 'intention', 'snapshot', 'neural_keyframe', 'neural_delta', 'contact', 'action_result', 'metrics', 'paused', 'error']);

function fail(message: string): never { throw new TypeError(`Invalid simulation protocol message: ${message}`); }
function isRecord(value: unknown): value is Record<string, unknown> { return typeof value === 'object' && value !== null && !Array.isArray(value); }
function finiteNumber(value: unknown, field: string, min: number, max: number): number {
  if (typeof value !== 'number' || !Number.isFinite(value) || value < min || value > max) fail(`${field} must be a finite number between ${min} and ${max}`);
  return value;
}
function integer(value: unknown, field: string, min: number, max: number): number {
  const number = finiteNumber(value, field, min, max);
  if (!Number.isSafeInteger(number)) fail(`${field} must be a safe integer`);
  return number;
}
function string(value: unknown, field: string, pattern?: RegExp): string {
  if (typeof value !== 'string' || value.length < 1 || value.length > MAX_STRING_LENGTH || (pattern && !pattern.test(value))) fail(`${field} is invalid`);
  return value;
}
function onlyKeys(value: Record<string, unknown>, keys: readonly string[]): void {
  const allowed = new Set(keys);
  for (const key of Object.keys(value)) if (!allowed.has(key)) fail(`unexpected key ${key}`);
}
function array(value: unknown, field: string, min: number, max: number): unknown[] {
  if (!Array.isArray(value) || value.length < min || value.length > max) fail(`${field} must contain between ${min} and ${max} entries`);
  return value;
}
function frame(value: unknown): Record<string, unknown> {
  let decoded = value;
  if (typeof decoded === 'string') {
    if (new TextEncoder().encode(decoded).byteLength > MAX_FRAME_BYTES) fail('JSON frame exceeds 1 MiB');
    try { decoded = JSON.parse(decoded); } catch { fail('JSON frame is malformed'); }
  }
  if (!isRecord(decoded)) fail('frame must be a JSON object');
  let encoded: string;
  try { encoded = JSON.stringify(decoded); } catch { fail('frame is not JSON serializable'); }
  if (encoded === undefined || new TextEncoder().encode(encoded).byteLength > MAX_FRAME_BYTES) fail('JSON frame exceeds 1 MiB');
  return decoded;
}
function envelope(value: Record<string, unknown>, keys: readonly string[]): void {
  onlyKeys(value, keys);
  if (value.version !== 2) fail('version must be 2');
  string(value.sessionId, 'sessionId', /^[se]-[a-z0-9]{8,64}$/);
  string(value.episodeId, 'episodeId', /^[se]-[a-z0-9]{8,64}$/);
  integer(value.sequence, 'sequence', 0, MAX_SEQUENCE);
  finiteNumber(value.simulationTime, 'simulationTime', 0, 86_400);
}
function validateUpdates(value: unknown): void {
  for (const update of array(value, 'updates', 0, MAX_NEURAL_UPDATES)) {
    if (!isRecord(update)) fail('update must be an object');
    onlyKeys(update, ['neuronId', 'value']);
    integer(update.neuronId, 'neuronId', 0, MAX_SEQUENCE);
    finiteNumber(update.value, 'value', -1_000_000, 1_000_000);
  }
}

/** Parse one strict, bounded server control frame. */
export function parseServerMessage(value: unknown): ServerMessage {
  const message = frame(value);
  if (typeof message.type !== 'string' || !serverTypes.has(message.type as ServerMessage['type'])) fail('type is not a supported server message');
  switch (message.type) {
    case 'ready': {
      envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime', 'backend', 'population', 'graphHash', 'checkpointHash', 'acceptedVersions', 'maxFrameBytes', 'maxNeuralUpdates']);
      if (!resolvedBackends.has(string(message.backend, 'backend'))) fail('backend is invalid');
      if (!populations.has(message.population as PopulationSize)) fail('population is invalid');
      string(message.graphHash, 'graphHash', /^[a-f0-9]{64}$/); string(message.checkpointHash, 'checkpointHash', /^[a-f0-9]{64}$/);
      if (message.acceptedVersions !== undefined) for (const version of array(message.acceptedVersions, 'acceptedVersions', 1, 1)) if (version !== 2) fail('acceptedVersions is invalid');
      if (message.maxFrameBytes !== undefined) integer(message.maxFrameBytes, 'maxFrameBytes', 1, MAX_FRAME_BYTES);
      if (message.maxNeuralUpdates !== undefined) integer(message.maxNeuralUpdates, 'maxNeuralUpdates', 1, MAX_NEURAL_UPDATES);
      break;
    }
    case 'reset_complete': envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime']); break;
    case 'intention':
      envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime', 'intentionId', 'action', 'motorPhase']);
      string(message.intentionId, 'intentionId', /^i-[a-z0-9]{8,64}$/); if (!actions.has(message.action as Action)) fail('action is invalid'); if (!motorPhases.has(message.motorPhase as MotorPhase)) fail('motorPhase is invalid'); break;
    case 'snapshot': {
      envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime', 'body', 'joints', 'keys']);
      for (const coordinate of array(message.body, 'body', 3, 3)) finiteNumber(coordinate, 'body', -10_000, 10_000);
      for (const joint of array(message.joints, 'joints', 0, 256)) finiteNumber(joint, 'joints', -100, 100);
      if (!isRecord(message.keys)) fail('keys must be an object');
      for (const [key, travel] of Object.entries(message.keys)) { if (!keyNames.has(key as KeyName)) fail('keys contains an invalid key'); finiteNumber(travel, 'keys', 0, 1); }
      break;
    }
    case 'neural_keyframe': case 'neural_delta':
      envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime', 'updates']); validateUpdates(message.updates); break;
    case 'contact':
      envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime', 'intentionId', 'requestedKey', 'touchedKey', 'travel', 'force', 'debounce', 'confirmed']);
      string(message.intentionId, 'intentionId', /^i-[a-z0-9]{8,64}$/); if (!keyNames.has(message.requestedKey as KeyName)) fail('requestedKey is invalid'); if (message.touchedKey !== null && !keyNames.has(message.touchedKey as KeyName)) fail('touchedKey is invalid'); finiteNumber(message.travel, 'travel', 0, 1); finiteNumber(message.force, 'force', 0, 10_000); finiteNumber(message.debounce, 'debounce', 0, 10); if (typeof message.confirmed !== 'boolean') fail('confirmed must be boolean'); break;
    case 'action_result':
      envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime', 'intentionId', 'result']); string(message.intentionId, 'intentionId', /^i-[a-z0-9]{8,64}$/); if (message.result !== 'confirmed' && message.result !== 'waited' && message.result !== 'failed') fail('result is invalid'); break;
    case 'metrics':
      envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime', 'physicsHz', 'motorHz', 'neuralHz', 'renderHz', 'latencyMs', 'droppedRenderFrames', 'backendUtilization']);
      for (const field of ['physicsHz', 'motorHz', 'neuralHz', 'renderHz'] as const) finiteNumber(message[field], field, 0, 100_000); finiteNumber(message.latencyMs, 'latencyMs', 0, 1_000_000); integer(message.droppedRenderFrames, 'droppedRenderFrames', 0, MAX_SEQUENCE); if (message.backendUtilization !== undefined) finiteNumber(message.backendUtilization, 'backendUtilization', 0, 1); break;
    case 'paused': case 'error':
      envelope(message, ['type', 'version', 'sessionId', 'episodeId', 'sequence', 'simulationTime', 'code', 'message']); string(message.code, 'code'); string(message.message, 'message'); break;
  }
  return message as ServerMessage;
}

File: python/fly_crossy/protocol.py
"""Strict, bounded wire schemas for simulation protocol version 2."""

from __future__ import annotations

import json
from collections.abc import Sequence
from typing import Annotated, Any, Literal, Self, TypeAlias

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    StrictFloat,
    StrictInt,
    StringConstraints,
    TypeAdapter,
    ValidationError,
    field_validator,
    model_validator,
)

MAX_FRAME_BYTES = 1_048_576
MAX_NEURAL_UPDATES = 20_000
MAX_NEURAL_CHUNKS = 256
MAX_STRING_LENGTH = 256
MAX_SEQUENCE = 2**53 - 1
MAX_SIMULATION_TIME = 86_400.0

PopulationSize: TypeAlias = Literal[80, 1000, 5000, 20000, 124289]
BackendPreference: TypeAlias = Literal["auto", "cpu", "gpu", "gpu-strict"]
ResolvedBackend: TypeAlias = Literal["cpu", "gpu"]
FallbackReason: TypeAlias = Literal["cuda-unavailable"]
Action: TypeAlias = Literal["forward", "backward", "left", "right", "wait"]
KeyName: TypeAlias = Literal["W", "A", "S", "D", "SPACE_LEFT", "SPACE_RIGHT"]
MotorPhase: TypeAlias = Literal[
    "neutral",
    "targeting",
    "reaching",
    "pressing",
    "confirmed",
    "lifting",
    "retracting",
    "settling",
    "failed",
]

ShortString = Annotated[str, StringConstraints(min_length=1, max_length=MAX_STRING_LENGTH)]
Hash = Annotated[str, StringConstraints(pattern=r"^[a-f0-9]{64}$")]
SessionOrEpisodeId = Annotated[str, StringConstraints(pattern=r"^[se]-[a-z0-9]{8,64}$")]
IntentionId = Annotated[str, StringConstraints(pattern=r"^i-[a-z0-9]{8,64}$")]
FiniteFloat = Annotated[StrictFloat, Field(allow_inf_nan=False)]
UnitFloat = Annotated[StrictFloat, Field(ge=0, le=1, allow_inf_nan=False)]
PositionFloat = Annotated[StrictFloat, Field(ge=-10_000, le=10_000, allow_inf_nan=False)]
JointFloat = Annotated[StrictFloat, Field(ge=-100, le=100, allow_inf_nan=False)]


class ProtocolModel(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
        validate_by_alias=True,
        validate_by_name=False,
        alias_generator=lambda name: "".join(
            part.capitalize() if index else part for index, part in enumerate(name.split("_"))
        ),
        allow_inf_nan=False,
    )


class Envelope(ProtocolModel):
    version: Literal[2]
    session_id: SessionOrEpisodeId
    episode_id: SessionOrEpisodeId
    sequence: Annotated[StrictInt, Field(ge=0, le=MAX_SEQUENCE)]
    simulation_time: Annotated[StrictFloat, Field(ge=0, le=MAX_SIMULATION_TIME, allow_inf_nan=False)]


class Hello(Envelope):
    type: Literal["hello"]
    supported_versions: Annotated[list[Literal[2]], Field(min_length=1, max_length=1)]
    ui_build: ShortString


class Configure(Envelope):
    type: Literal["configure"]
    population: PopulationSize
    backend: BackendPreference
    seed: Annotated[StrictInt, Field(ge=0, le=2**32 - 1)]
    speed: Annotated[StrictFloat, Field(gt=0, le=100, allow_inf_nan=False)]


class Reset(Envelope):
    type: Literal["reset"]
    seed: Annotated[StrictInt, Field(ge=0, le=2**32 - 1)] | None = None


class Observation(Envelope):
    type: Literal["observation"]
    game_step: Annotated[StrictInt, Field(ge=0, le=MAX_SEQUENCE)]
    observation: Annotated[list[FiniteFloat], Field(min_length=517, max_length=517)]
    reward: Annotated[StrictFloat, Field(ge=-1_000_000, le=1_000_000, allow_inf_nan=False)]


class Pause(Envelope):
    type: Literal["pause"]


class Resume(Envelope):
    type: Literal["resume"]


class RequestKeyframe(Envelope):
    type: Literal["request_keyframe"]


ClientMessage: TypeAlias = Hello | Configure | Reset | Observation | Pause | Resume | RequestKeyframe
ClientMessageAdapter = TypeAdapter(Annotated[ClientMessage, Field(discriminator="type")])


class Ready(Envelope):
    type: Literal["ready"]
    backend: ResolvedBackend
    population: PopulationSize
    graph_hash: Hash
    checkpoint_hash: Hash
    accepted_versions: Annotated[list[Literal[2]], Field(min_length=1, max_length=1)] | None = None
    max_frame_bytes: Annotated[StrictInt, Field(ge=1, le=MAX_FRAME_BYTES)] | None = None
    max_neural_updates: Annotated[StrictInt, Field(ge=1, le=MAX_NEURAL_UPDATES)] | None = None
    fallback_reason: FallbackReason | None = None

    @field_validator("fallback_reason", mode="before")
    @classmethod
    def reject_explicit_null_fallback_reason(cls, value: object) -> object:
        if value is None:
            raise ValueError("fallbackReason must be omitted or a supported reason")
        return value


class ResetComplete(Envelope):
    type: Literal["reset_complete"]


class Intention(Envelope):
    type: Literal["intention"]
    intention_id: IntentionId
    action: Action
    motor_phase: MotorPhase


class Snapshot(Envelope):
    type: Literal["snapshot"]
    body: Annotated[list[PositionFloat], Field(min_length=3, max_length=3)]
    joints: Annotated[list[JointFloat], Field(max_length=256)]
    keys: Annotated[dict[KeyName, UnitFloat], Field(max_length=6)]


class NeuralUpdate(ProtocolModel):
    neuron_id: Annotated[StrictInt, Field(ge=0, le=MAX_SEQUENCE)]
    value: Annotated[StrictFloat, Field(ge=-1_000_000, le=1_000_000, allow_inf_nan=False)]


class NeuralKeyframe(Envelope):
    type: Literal["neural_keyframe"]
    revision: Annotated[StrictInt, Field(ge=0, le=MAX_SEQUENCE)]
    chunk_index: Annotated[StrictInt, Field(ge=0, lt=MAX_NEURAL_CHUNKS)]
    chunk_count: Annotated[StrictInt, Field(ge=1, le=MAX_NEURAL_CHUNKS)]
    updates: Annotated[list[NeuralUpdate], Field(max_length=MAX_NEURAL_UPDATES)]

    @model_validator(mode="after")
    def validate_chunk_position(self) -> Self:
        if self.chunk_index >= self.chunk_count:
            raise ValueError("chunkIndex must be less than chunkCount")
        if any(update.value == 0 for update in self.updates):
            raise ValueError("neural keyframe updates must be nonzero")
        return self


class NeuralDelta(Envelope):
    type: Literal["neural_delta"]
    base_revision: Annotated[StrictInt, Field(ge=0, le=MAX_SEQUENCE)]
    revision: Annotated[StrictInt, Field(ge=1, le=MAX_SEQUENCE)]
    chunk_index: Annotated[StrictInt, Field(ge=0, lt=MAX_NEURAL_CHUNKS)]
    chunk_count: Annotated[StrictInt, Field(ge=1, le=MAX_NEURAL_CHUNKS)]
    updates: Annotated[list[NeuralUpdate], Field(max_length=MAX_NEURAL_UPDATES)]

    @model_validator(mode="after")
    def validate_revision_and_chunk_position(self) -> Self:
        if self.revision <= self.base_revision:
            raise ValueError("revision must be greater than baseRevision")
        if self.chunk_index >= self.chunk_count:
            raise ValueError("chunkIndex must be less than chunkCount")
        return self


class Contact(Envelope):
    type: Literal["contact"]
    intention_id: IntentionId
    requested_key: KeyName
    touched_key: KeyName | None
    travel: UnitFloat
    force: Annotated[StrictFloat, Field(ge=0, le=10_000, allow_inf_nan=False)]
    debounce: Annotated[StrictFloat, Field(ge=0, le=10, allow_inf_nan=False)]
    confirmed: bool


class ActionResult(Envelope):
    type: Literal["action_result"]
    intention_id: IntentionId
    result: Literal["confirmed", "waited", "failed"]


class Metrics(Envelope):
    type: Literal["metrics"]
    physics_hz: Annotated[StrictFloat, Field(ge=0, le=100_000, allow_inf_nan=False)]
    motor_hz: Annotated[StrictFloat, Field(ge=0, le=100_000, allow_inf_nan=False)]
    neural_hz: Annotated[StrictFloat, Field(ge=0, le=100_000, allow_inf_nan=False)]
    render_hz: Annotated[StrictFloat, Field(ge=0, le=100_000, allow_inf_nan=False)]
    latency_ms: Annotated[StrictFloat, Field(ge=0, le=1_000_000, allow_inf_nan=False)]
    dropped_render_frames: Annotated[StrictInt, Field(ge=0, le=MAX_SEQUENCE)]
    backend_utilization: UnitFloat | None = None


class Paused(Envelope):
    type: Literal["paused"]
    code: ShortString
    message: ShortString


class Error(Envelope):
    type: Literal["error"]
    code: ShortString
    message: ShortString


ServerMessage: TypeAlias = (
    Ready | ResetComplete | Intention | Snapshot | NeuralKeyframe | NeuralDelta | Contact | ActionResult | Metrics | Paused | Error
)
ServerMessageAdapter = TypeAdapter(Annotated[ServerMessage, Field(discriminator="type")])


def _load_frame(value: Any) -> Any:
    """Decode one bounded JSON control frame without accepting non-object payloads."""
    if isinstance(value, bytes):
        if len(value) > MAX_FRAME_BYTES:
            raise ValueError("Protocol JSON frame exceeds 1 MiB.")
        value = value.decode("utf-8")
    if isinstance(value, str):
        if len(value.encode("utf-8")) > MAX_FRAME_BYTES:
            raise ValueError("Protocol JSON frame exceeds 1 MiB.")
        value = json.loads(value)
    else:
        try:
            encoded = json.dumps(value, separators=(",", ":"), allow_nan=True).encode("utf-8")
        except (TypeError, ValueError) as error:
            raise ValueError("Protocol frame must be JSON serializable.") from error
        if len(encoded) > MAX_FRAME_BYTES:
            raise ValueError("Protocol JSON frame exceeds 1 MiB.")
    if not isinstance(value, dict):
        raise ValueError("Protocol frame must be a JSON object.")
    return value


def parse_client_message(value: Any) -> ClientMessage:
    """Validate a bounded v2 client message from decoded JSON or a JSON frame."""
    return ClientMessageAdapter.validate_python(_load_frame(value))


def parse_server_message(value: Any) -> ServerMessage:
    """Validate a bounded v2 server message from decoded JSON or a JSON frame."""
    return ServerMessageAdapter.validate_python(_load_frame(value))


def serialize_server_message(message: ServerMessage | BaseModel | dict[str, Any]) -> str:
    """Validate and serialize exactly one server frame within the advertised byte limit."""
    value = (
        message.model_dump(by_alias=True, exclude_none=True)
        if isinstance(message, BaseModel)
        else message
    )
    validated = ServerMessageAdapter.validate_python(value)
    payload = json.dumps(
        validated.model_dump(by_alias=True, exclude_none=True),
        separators=(",", ":"),
        allow_nan=False,
    )
    if len(payload.encode("utf-8")) > MAX_FRAME_BYTES:
        raise ValueError("Protocol JSON frame exceeds 1 MiB.")
    return payload


def chunk_neural_frames(
    *,
    message_type: Literal["neural_keyframe", "neural_delta"],
    session_id: str,
    episode_id: str,
    start_sequence: int,
    simulation_time: float,
    revision: int,
    updates: Sequence[NeuralUpdate | dict[str, Any]],
    base_revision: int | None = None,
) -> list[NeuralKeyframe | NeuralDelta]:
    """Build deterministic count- and byte-bounded neural wire chunks."""
    validated_updates = [
        update if isinstance(update, NeuralUpdate) else NeuralUpdate.model_validate(update)
        for update in updates
    ]
    parts = [
        validated_updates[index:index + MAX_NEURAL_UPDATES]
        for index in range(0, len(validated_updates), MAX_NEURAL_UPDATES)
    ] or [[]]

    while True:
        frames = [
            _neural_frame(
                message_type=message_type,
                session_id=session_id,
                episode_id=episode_id,
                sequence=start_sequence + chunk_index,
                simulation_time=simulation_time,
                base_revision=base_revision,
                revision=revision,
                chunk_index=chunk_index,
                chunk_count=len(parts),
                updates=part,
            )
            for chunk_index, part in enumerate(parts)
        ]
        oversized_index: int | None = None
        for index, frame in enumerate(frames):
            try:
                serialize_server_message(frame)
            except ValueError:
                if len(parts[index]) <= 1:
                    raise
                oversized_index = index
                break
        if oversized_index is None:
            return frames

        oversized = parts[oversized_index]
        midpoint = len(oversized) // 2
        parts[oversized_index:oversized_index + 1] = [
            oversized[:midpoint],
            oversized[midpoint:],
        ]


def _neural_frame(
    *,
    message_type: Literal["neural_keyframe", "neural_delta"],
    session_id: str,
    episode_id: str,
    sequence: int,
    simulation_time: float,
    revision: int,
    chunk_index: int,
    chunk_count: int,
    updates: list[NeuralUpdate],
    base_revision: int | None,
) -> NeuralKeyframe | NeuralDelta:
    value: dict[str, Any] = {
        "type": message_type,
        "version": 2,
        "sessionId": session_id,
        "episodeId": episode_id,
        "sequence": sequence,
        "simulationTime": simulation_time,
        "revision": revision,
        "chunkIndex": chunk_index,
        "chunkCount": chunk_count,
        "updates": [update.model_dump(by_alias=True) for update in updates],
    }
    if message_type == "neural_delta":
        if base_revision is None:
            raise ValueError("baseRevision is required for neural deltas")
        value["baseRevision"] = base_revision
        return NeuralDelta.model_validate(value)
    if base_revision is not None:
        raise ValueError("baseRevision is valid only for neural deltas")
    return NeuralKeyframe.model_validate(value)
